### ISSUE TITLE
Matches ranking and disambiguation

### ISSUE THREAD

**Issue by pombredanne** [nexB/purldb#240]
As we are starting to handle larger volumes of indexed packages, we are facing issues where there are multiple exact file matched for a given file. We eventually need to implement ranking and disambiguation of multiple matches to select the one best match among many. This means:
- [x] Create an updated matching pipeline that returns the best match
- [x] Create the function/procedure to disambiguate from multiple exact matches. This completed with https://github.com/nexB/purldb/issues/240#issuecomment-2032650241
- [x] https://github.com/nexB/purldb/issues/426

Note that this may also imply some changes or update to the indexing process and the data models in order to track the true origin of code. This will be tracked separately

This depends on #238

**Comment by pombredanne** [comment]
Exact matches ranking/disambiguation is done. (The remaining is to support ranking for approximate matches.)

- We established the pipeline by designing a sequence that goes from matching large aggregates to matching smaller files: whole archive, directories then files:
  - https://github.com/nexB/purldb/blob/ed53bcf9dfb1d0ea6e694efa3949556efd465cc7/matchcode_pipeline/pipelines/matching.py#L46

- We created a set of matching disambiguation code post processing the matching results in 
  - https://github.com/nexB/purldb/blob/ed53bcf9dfb1d0ea6e694efa3949556efd465cc7/matchcode_pipeline/pipes/matching.py#L330

- We did run extensive manual tests to review and refine the results of actual matching creating client-side disambiguation code at:
  - https://github.com/nexB/scancode.io/commit/45273a9afb4308d49b613b518ba2ba04a0dfdb30

Note that this is a feature that requires significant setup for local testing:
- You need to have a PurlDB, MatchCode.io and ScanCode.io installed locally.
Then you need to ensure that your PurlDB is populated with packages that contain some duplicated files and code.
- Then you need to create a project in the client ScanCode.io that will match some of its files to these duplicated packages, run a base scan and a purldb matching pipeline.
- The results are that you are seeing only the one good match of the multiple matches that exist in the PurlDB.

**Comment by pombredanne** [comment]
We now have a new MatchCode pipeline in  that returns the best match in https://github.com/nexB/purldb/blob/main/matchcode_pipeline/pipelines/matching.py

**Comment by JonoYang** [comment]
This has been merged into main. The `ApproximateMatchingHashMixin.match()` method has been updated to rank matches after the initial Hamming distance comparisons. The match ranking logic uses the Hamming distance, file size and name difference between the file we have. https://github.com/nexB/purldb/blob/main/matchcode/models.py#L348

All of these values are then used as sorting keys, where the matches whose sorting keys appear first are considered the best match (lowest difference).

Test instructions after installing and running PurlDB with it's accompanying ScanCode.io worker, and MatchCode.io:

    1. Index the package pkg:npm/deep-equal@1.0.1 using the api/collect/index_packages endpoint
    2. Using the api/matching/ endpoint, upload the test file https://github.com/nexB/purldb/blob/main/matchcode/tests/testfiles/match/approximate-file-matching/index-modified.js
    3. The package data for pkg:npm/deep-equal@1.0.1 should be present at api/matching/<uuid>/results
